### PR TITLE
Use DMLOptions to bypass duplicate rule exceptions

### LINKED ISSUE #1007: In an org with Duplicate Management set to 'allow', dlrs cannot update a duplicate Lead or Account. An exception is thrown.
**Is your feature request related to a problem? Please describe.**
The rollup field is not updated and Apex exception emails go out

**Describe the solution you'd like**
DML Options can be used on the update to allow the update to succeed.
[DMLOptions.DuplicateRuleHeader Class](https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_class_Database_DMLOptions_DuplicateRuleHeader.htm#apex_class_Database_DMLOptions_DuplicateRuleHeader
or the SObject.setOptions(DMLOptions) method.
It would be good for the update to happen and the rollups on duplicate records to be correct. I can't think of any reason not to use this DML Option, unless it affects sObjects not under duplicate management.

**Describe alternatives you've considered**
none

**Additional context**
Perhaps it is possible to make the DML Options part of the rollup setup screen so users can decide if they want the exception or an update to happen.

### PR DESCRIPTION
**Pull request by chininchu**
# Critical Changes

- None

# Changes

- Implemented DMLOptions to allow DLRS updates on duplicate Leads and Accounts
- Modified DLRS update logic to bypass duplicate rule exceptions
- Updated rollup calculation process to correctly handle duplicate records

# Issues Closed

- Fixes #1007: DLRS unable to update duplicate Leads/Accounts when Duplicate Management is set to 'allow'

### CODE REVIEW

**Comment by chininchu** [comment]
@aheber 
Ready for review

**Comment by aheber** [comment]
@chininchu thanks for getting this in! It looks like you've discarded the "allOrNothing" flag. You should be able to add it to the DMLOptions instance and retest.

Could you revert the changes to the other files? Feel free to submit them separately. The change to update Prettier version would be a great additional PR.

File: dlrs/main/classes/RollupService.cls [hunk 1]
```diff
@@ -1661,8 +1661,21 @@ global with sharing class RollupService {
...
+      // Create DmlOptions instance
+
+      Database.DmlOptions dmlOptions = new Database.DmlOptions();
+
+      // Allows save even if duplicated are detected
+      dmlOptions.duplicateRuleHeader.allowSave = true;
+
+      // Bypass duplicate rules
+
+      // dmlOptions.duplicateRuleHeader.duplicateRuleBypassFlow = true;
```
Comment (aheber) [inline dlrs/main/classes/RollupService.cls:1674]: Could you remove this extra comment?

**Review by aheber — CHANGES_REQUESTED** [review]

File: dlrs/main/classes/RollupService.cls [hunk 2]
```diff
@@ -1661,8 +1661,21 @@ global with sharing class RollupService {
...
+      Database.DmlOptions dmlOptions = new Database.DmlOptions();
+
+      // Allows save even if duplicated are detected
+      dmlOptions.duplicateRuleHeader.allowSave = true;
+
+      // Bypass duplicate rules
+
+
       try {
         return Database.update(masterRecords, allOrNothing);
```
Comment (aheber) [inline dlrs/main/classes/RollupService.cls:1676]: @chininchu were you able to test this code in an environment? You've put it back the way it was and aren't including the DmlOptions anymore.

File: dlrs/main/classes/RollupService.cls [hunk 3]
```diff
@@ -1661,8 +1661,21 @@ global with sharing class RollupService {
...
+
```
Comment (aheber) [inline dlrs/main/classes/RollupService.cls:1677]: Are these extra lines necessary?

File: dlrs/main/lwc/jsconfig.json [hunk 4]
```diff
@@ -1,12 +1,25 @@
...
-    "experimentalDecorators": true
-  },
-  "include": ["**/*", "../../../.sfdx/typings/lwc/**/*.d.ts"],
-  "paths": {
-    "c/*": ["*"]
-  },
-  "typeAcquisition": {
-    "include": ["jest"]
-  }
+    "compilerOptions": {
```
Comment (aheber) [inline dlrs/main/lwc/jsconfig.json:2]: This file is still showing modified in your PR. Please revert this change.

File: package-lock.json [hunk 5]
```diff
@@ -25,8 +25,8 @@
         "eslint-plugin-jest": "^24.7.0",
         "husky": "^6.0.0",
         "lint-staged": "^11.2.6",
-        "prettier": "^2.5.0",
-        "prettier-plugin-apex": "^1.10.0"
+        "prettier": "2.5.0",
```
Comment (aheber) [inline package-lock.json:28]: This file is still showing modified in your PR. Please revert this change.

**Comment by chininchu** [comment]
> @chininchu thanks for getting this in! It looks like you've discarded the "allOrNothing" flag. You should be able to add it to the DMLOptions instance and retest.
> 
> Could you revert the changes to the other files? Feel free to submit them separately. The change to update Prettier version would be a great additional PR.

@aheber, Please refer to my latest commit for changes. Thanks.

File: dlrs/main/classes/RollupService.cls (hunk 2 above)
Comment (chininchu) [reply dlrs/main/classes/RollupService.cls:1675]: Yes, I was able to test the code and it works. You should now see the DML options on my latest commit.

File: dlrs/main/lwc/jsconfig.json (hunk 4 above)
Comment (chininchu) [reply dlrs/main/lwc/jsconfig.json:2]: This should be aligned with what's on the main branch.

File: package-lock.json (hunk 5 above)
Comment (chininchu) [reply package-lock.json:28]: This should be aligned with what's on the main branch.

**Review by coriobriensfdo — APPROVED** [review]

**Comment by aheber** [comment]
Superseded by #1502
